crypto/bn/bn_lib.c

```diff
@@ -181,7 +181,7 @@ void BN_clear_free(BIGNUM *a)
 {
     int i;
 
-    if (a == NULL)
+    if (a == NULL || a->flags & BN_FLG_RDONLY)
         return;
     bn_check_top(a);
     if (a->d != NULL) {
```

Comment:
Let's consider lines next to this one. It's OPENSSL_cleanse followed by check for STATIC_DATA and conditional free. It's should be noted however that most (if not all) `a->d` vectors in STATIC_DATA structures reside in read-only segment. So that OPENSSL_cleanse is bound to crash with SEGV in such cases. And it does, just try BN_clear_free(BN_value_one()). The fact that we haven't ever observed it ought to mean that nobody calls it with static data. This in turn ought to mean that it would be appropriate to move OPENSSL_cleanse inside the condition. If we move even further, we'll find ourselves making exactly same argument, i.e. since even most (if not all) STATIC_DATA `*a` themselves residing in read-only segment, so that even second OPENSSL_cleanse should be conditional.  But in this case it becomes equivalent to RDONLY. Another reason for skepticism toward introduction of RDONLY is following. There are more cases where STATIC_DATA is meaningful, e.g. bn_expand_internal, so question is shouldn't RDONLY get same treatment? And when you tie all things up, what would be the difference between the two?

Reply:
Well I'm happy to remove RDONLY and make BN_clear_free() not cleanse a->d if STATIC_DATA is set and to not cleanse *a if BN_FLG_MALLOCED is not set either. Now that we've made BIGNUM opaque applications can't meaningfully create a static BIGNUM with BN_FLG_MALLOCED clear.

The aim of that commit is so DH_free() doesn't crash when it tries to free up read only BIGNUM structures. It needlessly calls BN_clear_free() on various parameters. Changing it to BN_free() doesn't help though because BN_free() attempts to modify a a BIGNUM further and uses the mysterious BN_FLG_FREE flag which AFAICS has never been used by anything.